src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
+        self.deprecate_deltas_before(self.serial + 1);
+
+        self.serial = notification_file.serial();
+        self.session_id = notification_file.session_id(); // could have been reset.
+
+        self.current_objects =
+            CurrentObjectMap::read_snapshot(notification_file.snapshot(), fetcher)?;
+        self.snapshot = self
+            .current_objects
+            .derive_snapshot(self.session_id(), self.serial());
+        self.deltas = Self::read_deltas(notification_file.deltas(), fetcher)?;
 
-    fn meta(&mut self, session_id: Uuid, serial: u64) -> Result<(), Self::Err> {
-        self.meta = Some(ParsedMeta { session_id, serial });
         Ok(())
     }
 
-    fn publish(&mut self, uri: uri::Rsync, data: Vec<u8>) -> Result<(), Self::Err> {
-        self.publish.insert(ComparableRsyncUri(uri), data);
+    /// Write out all *missing* RRDP files. Optionally delay writing the notification file for
+    /// the specified number of seconds
+    pub fn write_rrdp_files(&self, notify_delay: u64) -> Result<()> {
+        self.write_snapshot()?;
+        self.write_missing_deltas()?;
+
+        if notify_delay > 0 {
+            info!(
+                "Waiting {} seconds before writing RRDP notification file",
+                notify_delay
+            );
+            std::thread::sleep(std::time::Duration::from_secs(notify_delay));
+        }
+        self.write_notification()?;
+
         Ok(())
     }
-}
 
-fn parse_snapshot(buf: &[u8]) -> Result<SnapshotParser> {
-    let cursor = std::io::Cursor::new(buf);
-    let reader = BufReader::new(cursor);
-    let mut processor = SnapshotParser::new();
-    processor.process(reader).map_err(|err| {
-        anyhow!("Error parsing snapshot XML: {:?}", &err)
-    })?;
-    Ok(processor)
-}
+    /// Cleans deprecated files and their parent directories if they are empty
+    pub fn clean(&mut self, config: &Config) -> Result<()> {
+        let clean_before = Time::seconds_ago(config.cleanup_after);
+
+        for deprecated in self
+            .deprecated_files
+            .iter()
+            .filter(|d| d.since < clean_before)
+        {
+            let path = &deprecated.path;
+            if path.exists() {
+                info!(
+                    "Removing RRDP file: {:?}, deprecated since: {}",
+                    path, deprecated.since
+                );
+                file_ops::remove_file_and_empty_parent_dirs(path)?;
+            }
+        }
 
-// DeltaParser parses RRDP delta XML and adds/removes records to/from the
-// referenced snapshot for publish/withdraw delta records respectively. As the
-// DeltaParser doesn't own the snapshot it needs a mutable reference to it which
-// is guaranteed (by 'a) to be valid as long as the DeltaParser instance exists.
-struct DeltaParser<'a> {
-    pub meta: Option<ParsedMeta>,
-    pub snapshot: &'a mut SnapshotParser,
-}
+        self.deprecated_files.retain(|d| d.since > clean_before);
```

Comment:
Same question about < and > but not == as I mentioned in a comment elsewhere already. Also, if this code is so similar to that other code, perhaps merge them? (I don't have time to go back and find that code now and see if this suggestion really makes sense or not, sorry).

Reply:
Fixed. Not easy to merge - subtle differences like log messages and removing files and optionally empty parent dirs for RRDP vs complete rsync dirs for rsync.